File: javascript/src/validator.ts
import { timingSafeEqual } from 'crypto';
import { generateSignature } from './signer';
import { HEADERS, DEFAULT_TOLERANCE_SECONDS } from './constants';
import {
  MissingHeadersError,
  InvalidSignatureError,
  InvalidTimestampError,
  TimestampExpiredError,
} from './errors';

export type WebhookHeaders = Record<string, string | string[] | undefined> | Map<string, string>;

export interface WebhookValidatorOptions {
  secretKey: string;
  toleranceInSeconds?: number;
}

export class WebhookValidator {
  private readonly secretKey: string;
  private readonly toleranceMs: number;

  constructor(options: WebhookValidatorOptions) {
    const { secretKey, toleranceInSeconds = DEFAULT_TOLERANCE_SECONDS } = options;

    if (!secretKey || secretKey.trim() === '') {
      throw new Error('secretKey cannot be empty');
    }

    this.secretKey = secretKey;
    this.toleranceMs = toleranceInSeconds * 1000;
  }

  verify<T = unknown>(payload: string, headers: WebhookHeaders): T {
    const id = this.getHeader(headers, HEADERS.ID);
    const timestampStr = this.getHeader(headers, HEADERS.TIMESTAMP);
    const signature = this.getHeader(headers, HEADERS.SIGNATURE);

    const missingHeaders: string[] = [];
    if (!id) missingHeaders.push(HEADERS.ID);
    if (!timestampStr) missingHeaders.push(HEADERS.TIMESTAMP);
    if (!signature) missingHeaders.push(HEADERS.SIGNATURE);

    if (missingHeaders.length > 0) {
      throw new MissingHeadersError(missingHeaders);
    }

    const timestampMs = parseInt(timestampStr!, 10);
    if (isNaN(timestampMs)) {
      throw new InvalidTimestampError();
    }

    const currentMs = Date.now();
    const diff = Math.abs(currentMs - timestampMs);

    if (diff > this.toleranceMs) {
      throw new TimestampExpiredError();
    }

    const signaturePayload = `${id}.${timestampMs}.${payload}`;

    const expectedSignature = generateSignature(this.secretKey, signaturePayload);

    if (!this.constantTimeCompare(expectedSignature, signature!)) {
      throw new InvalidSignatureError();
    }

    return JSON.parse(payload) as T;
  }

  private getHeader(headers: WebhookHeaders, name: string): string | undefined {
    if (headers instanceof Map) {
      return headers.get(name);
    }

    const value = headers[name] ?? headers[name.toLowerCase()];
    if (Array.isArray(value)) {
      return value[0];
    }
    return value;
  }

  private constantTimeCompare(a: string, b: string): boolean {
    try {
      const bufA = Buffer.from(a, 'base64');
      const bufB = Buffer.from(b, 'base64');

      if (bufA.length !== bufB.length) {
        return false;
      }

      return timingSafeEqual(bufA, bufB);
    } catch {
      return false;
    }
  }
}

File: go/validator.go
package webhook

import (
	"crypto/hmac"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
	"time"
)

type WebhookValidator struct {
	secretKey   string
	toleranceMs int64
}

func NewWebhookValidator(secretKey string) *WebhookValidator {
	return NewWebhookValidatorWithTolerance(secretKey, DefaultToleranceSeconds)
}

func NewWebhookValidatorWithTolerance(secretKey string, toleranceSeconds int64) *WebhookValidator {
	if strings.TrimSpace(secretKey) == "" {
		panic("secretKey cannot be empty")
	}

	return &WebhookValidator{
		secretKey:   secretKey,
		toleranceMs: toleranceSeconds * 1000,
	}
}

func (v *WebhookValidator) Verify(payload string, headers map[string]string) (map[string]interface{}, error) {
	var result map[string]interface{}
	err := v.VerifyAndUnmarshal(payload, headers, &result)
	return result, err
}

func (v *WebhookValidator) VerifyAndUnmarshal(payload string, headers map[string]string, target interface{}) error {
	var missingHeaders []string

	id, hasID := headers[HeaderID]
	timestampStr, hasTimestamp := headers[HeaderTimestamp]
	signature, hasSignature := headers[HeaderSignature]

	if !hasID || id == "" {
		missingHeaders = append(missingHeaders, HeaderID)
	}
	if !hasTimestamp || timestampStr == "" {
		missingHeaders = append(missingHeaders, HeaderTimestamp)
	}
	if !hasSignature || signature == "" {
		missingHeaders = append(missingHeaders, HeaderSignature)
	}

	if len(missingHeaders) > 0 {
		return &MissingHeadersError{Headers: missingHeaders}
	}

	timestampMs, err := strconv.ParseInt(timestampStr, 10, 64)
	if err != nil {
		return &InvalidTimestampError{}
	}

	currentMs := time.Now().UnixMilli()
	diff := currentMs - timestampMs
	if diff < 0 {
		diff = -diff
	}

	if diff > v.toleranceMs {
		return &TimestampExpiredError{}
	}

	signaturePayload := fmt.Sprintf("%s.%d.%s", id, timestampMs, payload)

	expectedSignature := GenerateSignature(v.secretKey, signaturePayload)

	if !constantTimeCompare(expectedSignature, signature) {
		return &InvalidSignatureError{}
	}

	return json.Unmarshal([]byte(payload), target)
}

func constantTimeCompare(a, b string) bool {
	aBytes, errA := base64.StdEncoding.DecodeString(a)
	bBytes, errB := base64.StdEncoding.DecodeString(b)

	if errA != nil || errB != nil {
		return false
	}

	return hmac.Equal(aBytes, bBytes)
}
